registerApiRoute createHandler is typed (c: Context) but called with { mastra }

## Describe the bug

`createHandler` in `registerApiRoute` is typed to take a Hono `Context`, but at runtime it's
called with `{ mastra }`. So the way you're meant to use it (grab `mastra` for one-time setup)
doesn't typecheck.

The public option type and the type the package actually produces disagree, and only the second
one matches what runs:

`packages/core/src/server/index.ts:56` (the public option type):

```ts
createHandler?: (c: Context) => Promise<ApiRouteHandler>;
```

`packages/core/src/server/types.ts:36` (the `ApiRoute` it produces):

```ts
createHandler: ({ mastra }: { mastra: Mastra }) => Promise<ApiRouteHandler>;
```

`packages/server/src/server/server-adapter/index.ts:904` (what actually calls it):

```ts
await route.createHandler({ mastra })
```

So in practice you either follow the type and write `(c) => c.get('mastra')`, which throws at
runtime because `c` is just `{ mastra }` and has no `.get`/`.req`, or you write the correct
`({ mastra }) => ...` and cast to keep the compiler quiet.

The `(c: any)` on the returned handler is deliberate (there's a type test for it), so that part
isn't what I'm reporting. This is only about the factory argument.

## Steps to reproduce

```ts
import { registerApiRoute } from '@mastra/core/server';

registerApiRoute('/x', {
  method: 'GET',
  createHandler: async ({ mastra }) => { // TS2339: Property 'mastra' does not exist on type 'Context<...>'
    const agent = mastra.getAgent('weather');
    return async c => c.json({ ok: true });
  },
});
```

On `@mastra/core@1.50.1`, `tsc --noEmit` gives:

```
src/index.ts(9,27): error TS2339: Property 'mastra' does not exist on type 'Context<any, any, {}>'.
```

The existing createHandler type test calls the factory as `async () => ...` (no argument), so it
doesn't catch this.

## Link to minimal reproducible example

https://github.com/tehfreak/repro-mastra-createhandler-arg-type

`npm install && npm run check`. There's also a GitHub Actions run showing the same error in a
clean environment.

## Suggested fix

```diff
- createHandler?: (c: Context) => Promise<ApiRouteHandler>;
+ createHandler?: (opts: { mastra: Mastra }) => Promise<ApiRouteHandler>;
```

Happy to open a PR for this.

## Environment

`@mastra/core@1.50.1` (current latest on npm), also on `main`.

--- a/.changeset/fix-createhandler-arg-type.md
+++ b/.changeset/fix-createhandler-arg-type.md
@@ -1,0 +1,5 @@
+---
+'@mastra/core': patch
+---
+
+Fix the `createHandler` option type on `registerApiRoute`. It's called with `{ mastra }` at runtime but was typed as `(c: Context)`; it's now `(opts: { mastra: Mastra }) => Promise<ApiRouteHandler>`, matching the runtime and the internal `ApiRoute` type.

--- a/docs/src/content/en/reference/server/register-api-route.mdx
+++ b/docs/src/content/en/reference/server/register-api-route.mdx
@@ -47,8 +47,9 @@ registerApiRoute("/items/:itemId", { ... })
   },
   {
     name: 'createHandler',
-    type: '(c: Context) => Promise<Handler>',
-    description: 'Async factory function to create the handler. Use either `handler` or `createHandler`, not both.',
+    type: '({ mastra }: { mastra: Mastra }) => Promise<ApiRouteHandler>',
+    description:
+      'Async factory that receives the Mastra instance and returns the route handler. Runs once at server startup, so it can perform one-time setup. Use either `handler` or `createHandler`, not both.',
     isOptional: true,
   },
   {
@@ -303,12 +304,13 @@ For routes that need async initialization:
 ```typescript
 registerApiRoute('/dynamic', {
   method: 'GET',
-  createHandler: async c => {
-    // Perform async setup
+  createHandler: async ({ mastra }) => {
+    // Perform one-time async setup
     const config = await loadConfig()
+    const agent = mastra.getAgent('weatherAgent')
 
     return async c => {
-      return c.json({ config })
+      return c.json({ config, agent: agent.name })
     }
   },
 })

--- a/packages/core/src/server/index.ts
+++ b/packages/core/src/server/index.ts
@@ -1,4 +1,4 @@
-import type { Context, Handler, MiddlewareHandler } from 'hono';
+import type { Handler, MiddlewareHandler } from 'hono';
 import type { DescribeRouteOptions } from 'hono-openapi';
 import { MastraError, ErrorDomain, ErrorCategory } from '../error';
 import type { Mastra } from '../mastra';
@@ -53,7 +53,7 @@ type RegisterApiRouteOptions<P extends string> = {
     P,
     ParamsFromPath<P>
   >;
-  createHandler?: (c: Context) => Promise<ApiRouteHandler>;
+  createHandler?: (opts: { mastra: Mastra }) => Promise<ApiRouteHandler>;
   middleware?: MiddlewareHandler | MiddlewareHandler[];
   /**
    * Route-specific CORS configuration.

--- a/packages/core/src/server/server.test-d.ts
+++ b/packages/core/src/server/server.test-d.ts
@@ -64,6 +64,19 @@ describe('registerApiRoute Type Tests', () => {
         },
       });
     });
+
+    it('types the createHandler factory arg as { mastra }, not a Hono Context', () => {
+      registerApiRoute('/factory', {
+        method: 'GET',
+        createHandler: async ({ mastra }) => {
+          // adapter calls this as route.createHandler({ mastra }), so mastra
+          // should be the Mastra instance, not a Hono Context
+          expectTypeOf(mastra).toEqualTypeOf<Mastra>();
+
+          return async c => c.json({ ok: true });
+        },
+      });
+    });
   });
 });
 

